keras_tuner/tuners/genetic.py

```diff
@@ -0,0 +1,542 @@
+        self._seed_state = self.seed
+        self._random_state = np.random.RandomState(self.seed)
+
+        if self.mutation_factor + self.crossover_factor > 1.0:
+            raise ValueError(
+                "The sum of the 'mutation_factors' and 'crossover_factors' "
+                "must be less than 1.0."
+            )
+        if self.selection_type not in ["roulette_wheel", "tournament"]:
+            raise ValueError(
+                "The 'selection_type' must be either 'roulette_wheel' or "
+                "'tournament'."
+            )
+        self._tried_so_far = set()
+        self._max_collisions = 20
+        self.gep = self._make_gep()
+        self.ranges = self._make_ranges
+        self.population = {"hyperparameters": [], "scores": []}
+        self.new_population = {"hyperparameters": [], "scores": []}
+        self.values = {hp.name: hp.default for hp in self.get_space().space}
+        self.parent_1, self.parent_2 = None, None
+
+    @property
+    def _make_max_trials(self):
+        """Calculate the maximum number of trials."""
+        return (
+            self.population_size
+            + self.generation_size * 2 * self.offspring_size
+        )
+
+    @property
+    def _make_ranges(self):
+        """Make the ranges for genetic optimization with respect to max trial.
+
+        Due to the fact that 'oracle' refers to the 'max_trial',
+        the ranges for the genetic optimization have to be calculated
+        for:
+
+            1.  'population_range': the number of models in the
+                population.
+            2.  'offspring_ranges': the number of offspring to produce
+                at each generation.
+            3. 'second_parent_range': the number of models to select
+            4.  'generation_range': the number of generations, where the
+                best models are selected and the offspring are produced.
+
+        Note, 'second_parent_range' is part of the 'offspring_ranges',
+        however, the second evaluation of the 'offspring_ranges' has
+        to be calculated separately, because only one value can be
+        returned once.
+
+        Returns:
+            A dict of four lists: 'population_range', 'generation_range',
+             'offspring_range', 'second_parent_range'.
+        """
+
+        population_range = list(range(self.population_size))
+        generation_range = list(
+            range(
+                self.population_size + self.offspring_size * 2,
+                self.max_trials,
+                self.offspring_size * 2 - 1,
+            )
+        )
+        offspring_range = list(range(self.population_size, self.max_trials))
+        first_parent_range = offspring_range[::2]
+        second_parent_range = offspring_range[1::2]
+        return {
+            "population": population_range,
+            "generation": generation_range,
+            "first_parent": first_parent_range,
+            "second_parent": second_parent_range,
+        }
```

Comment:
No return value is needed. Just use `self.population_range` and `self.generation_range` and so on.